src/engine/Union.cpp

```diff
@@ -292,3 +334,185 @@ std::unique_ptr<Operation> Union::cloneImpl() const {
   }
   return copy;
 }
+
+// _____________________________________________________________________________
+std::shared_ptr<Operation> Union::createSortedVariant(
+    const vector<ColumnIndex>& sortOrder) const {
+  return std::make_shared<Union>(_executionContext, _subtrees.at(0),
+                                 _subtrees.at(1), sortOrder);
+}
+
+namespace {
+struct Wrapper {
+  const IdTable& idTable_;
+  const LocalVocab& localVocab_;
+};
+Result::IdTableVocabPair& moveOrCopy(Result::IdTableVocabPair& element) {
+  return element;
+}
+Result::IdTableVocabPair moveOrCopy(const Wrapper& element) {
+  return {element.idTable_.clone(), element.localVocab_.clone()};
+}
+}  // namespace
+
+// _____________________________________________________________________________
+// Always inline makes makes a huge difference on large datasets.
+template <size_t SPAN_SIZE>
+AD_ALWAYS_INLINE bool Union::isSmaller(const auto& row1,
+                                       const auto& row2) const {
+  using StaticRange = std::span<const ColumnIndex, SPAN_SIZE>;
+  for (auto& col : StaticRange{targetOrder_}) {
+    ColumnIndex index1 = _columnOrigins.at(col).at(0);
+    ColumnIndex index2 = _columnOrigins.at(col).at(1);
```

Comment:
` auto [index1, index2] = _columnOrigins.at(col)`
And measure, whether the check in the at() function costs something, and 
whether we get a speedup by reordering the columns, such that they are alwasy 0, 1, ... in both inputs and the lookup in ` _columnOrigins` in each loop disappears
(if you want to, first create a separate benchmark, if you don't want to do all the permutation reordering first, although it is rather simple.